mypy/checker.py

```diff
@@ -1901,8 +1901,20 @@ def check_assignment(self, lvalue: Lvalue, rvalue: Expression, infer_lvalue_type
                 rvalue_type = self.expr_checker.accept(rvalue)
                 if not inferred.is_final:
                     rvalue_type = remove_instance_last_known_values(rvalue_type)
+                self.check_forbidden_inference_types(rvalue_type, rvalue)
```

Comment:
I think this should be triggered only if `rvalue` is a list or dict literal, like `[x, y, z]` or `{1: 2, 3: 4}`.